exchanger/exchanger.go

```diff
@@ -70,50 +51,18 @@ func ErrorLogging(l *log.Logger) Decorator {
 }
 
 // Instrumentation returns a Decorator which instruments an Exchanger with the given
-// counter.
-func Instrumentation(c logging.Counter) Decorator {
-	return func(ex Exchanger) Exchanger {
-		return Func(func(m *dns.Msg, a string) (*dns.Msg, time.Duration, error) {
-			defer c.Inc()
-			return ex.Exchange(m, a)
-		})
-	}
-}
-
-// A Recurser returns the addr (host:port) of the next DNS server to recurse a
-// Msg to. Empty returns signal that further recursion isn't possible or needed.
-type Recurser func(*dns.Msg) string
-
-// Recurse is the default Mesos-DNS Recurser which returns an addr (host:port)
-// only when the given dns.Msg doesn't contain authoritative answers and has at
-// least one SOA record in its NS section.
-func Recurse(r *dns.Msg) string {
-	if r.Authoritative && len(r.Answer) > 0 {
-		return ""
-	}
-
-	for _, ns := range r.Ns {
-		if soa, ok := ns.(*dns.SOA); ok {
-			return net.JoinHostPort(soa.Ns, "53")
-		}
-	}
-
-	return ""
-}
-
-// Recursion returns a Decorator which recurses until the given Recurser returns
-// an empty string or max attempts have been reached.
-func Recursion(max int, rec Recurser) Decorator {
+// counters.
+func Instrumentation(total, success, failure logging.Counter) Decorator {
 	return func(ex Exchanger) Exchanger {
 		return Func(func(m *dns.Msg, a string) (r *dns.Msg, rtt time.Duration, err error) {
-			for i := 0; i <= max; i++ {
-				if r, rtt, err = ex.Exchange(m, a); err != nil {
-					break
-				} else if a = rec(r); a == "" {
-					break
+			defer func() {
+				if total.Inc(); err != nil {
```

Comment:
this looks strange because the increment of success depends on a negative outcome of total.Inc(). unless I'm missing something, total.Inc() should really be on its own line.